pkg/roachprod/vm/ibm/helpers.go

```diff
@@ -0,0 +1,1255 @@
+			continue
+		}
+
+		if instanceResp.PrimaryNetworkAttachment == nil ||
+			instanceResp.PrimaryNetworkAttachment.VirtualNetworkInterface == nil {
+			continue
+		}
+
+		// Fill missing boot volume information.
+		if instanceResp.BootVolumeAttachment != nil && instanceResp.BootVolumeAttachment.Volume != nil {
+			err := bootVolume.updateWithID(
+				*instanceResp.BootVolumeAttachment.Volume.ID,
+				*instanceResp.BootVolumeAttachment.Volume.CRN,
+			)
+			if err != nil {
+				return nil, errors.Wrapf(
+					err,
+					"failed to update boot volume %s with ID %s",
+					*instanceResp.BootVolumeAttachment.Volume.Name,
+					*instanceResp.BootVolumeAttachment.Volume.ID,
+				)
+			}
+		}
+
+		// Fill missing attached volumes information.
+		for _, attachedVolume := range instanceResp.VolumeAttachments {
+			if attachedVolume.Volume == nil {
+				continue
+			}
+
+			if *attachedVolume.Volume.ID == bootVolume.ID {
+				continue
+			}
+
+			err := attachedVolumes.updateWithIDByName(
+				*attachedVolume.Volume.Name,
+				*attachedVolume.Volume.ID,
+				*attachedVolume.Volume.CRN,
+			)
+			if err != nil {
+				return nil, errors.Wrapf(
+					err,
+					"failed to update attached volume %s with ID %s",
+					*attachedVolume.Volume.Name,
+					*attachedVolume.Volume.ID,
+				)
+			}
+		}
+
+		instanceReady = true
+		break
+	}
+
+	if !instanceReady {
+		return nil, errors.New("instance did not reach running state in time")
+	}
+
+	// Create and attach a floating IP address to the instance
+	fipResp, err := p.attachOrCreateFloatingIPAddress(
+		l,
+		opts.zone,
+		p.config.roachprodResourceGroupID,
+		*instanceResp.ID,
+		*instanceResp.PrimaryNetworkAttachment.VirtualNetworkInterface.ID,
+		0,
+	)
+	if err != nil {
+		return nil, errors.Wrapf(
+			err,
+			"failed to create floating IP for instance %s in zone %s",
+			opts.vmName,
+			opts.zone,
+		)
+	}
+
+	// Create an instance object with all the necessary information.
+	vm := (&instance{
+		provider:         p,
+		instance:         instanceResp,
+		bootVolume:       bootVolume,
+		attachedVolumes:  attachedVolumes,
+		networkInterface: (&vpcv1FloatingIP{fipResp}).toVpcv1NetworkInterface(),
+		tagList:          &opts.tags,
+	}).toVM()
+
+	return &vm, nil
+}
+
+func (p *Provider) attachOrCreateFloatingIPAddress(
+	l *logger.Logger, zone, resourceGroupID, instanceID, networkInterfaceID string, attempt int,
+) (*vpcv1.FloatingIP, error) {
+
+	if attempt >= maxFloatingIPAttempts {
+		return nil, errors.Errorf(
+			"failed to create floating IP address after %d attempts",
+			maxFloatingIPAttempts,
+		)
+	}
+
+	l.Printf(
+		"Floating IP requested for instance %s in zone %s (attempt %d)",
+		instanceID,
+		zone,
+		attempt+1,
+	)
+
+	vpcService, err := p.getVpcServiceFromZone(zone)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to get VPC service for region")
+	}
+
+	// List floating IP addresses
+	floatingIPAddressesPager, err := vpcService.NewFloatingIpsPager(
```

Comment:
This could be useful for troubleshooting via CLI. Maybe add a TODO to refactor into a separate helper which can be invoked via CLI, directly.